transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -457,6 +458,24 @@ public ChannelFuture deregister() {
         return deregister(newPromise());
     }
 
+    /**
+     * If possible check if the given {@link ChannelPromise} is using the same {@link EventExecutor} as this
+     * {@link ChannelHandlerContext} and if not return a new {@link ChannelPromise} that runs on the same
+     * {@link EventExecutor} as this {@link ChannelHandlerContext}. The result of the new {@link ChannelPromise} is
+     * cascaded to the old {@link ChannelPromise}.
+     *
+     * This is done to ensure that {@link ChannelFutureListener}s that are added to the {@link ChannelPromise} by an
+     * {@link ChannelOutboundHandler} are executed in the same thread as the handler itself. By doing so we can
+     * ensure that there are not issues even if fields etc that are stored in the handler are modified by the listener.
+     */
+    private ChannelPromise ensurePromiseUseCorrectExecutor(ChannelPromise promise) {
+        if (promise instanceof DefaultChannelPromise && !((DefaultChannelPromise) promise).executor().inEventLoop()) {
```

Comment:
if we merge https://github.com/netty/netty/pull/16039 we can do this without a cast in 5.0 and so make it work for any `Promise` implementation.